source/core/module_coreHelpers.js

```diff
@@ -75,33 +104,68 @@
                     str += String.fromCharCode(u0);
                     continue;
                 }
+
+                // Look ahead to validate the UTF-8 structure
+                if ((u0 & 0xE0) === 0xC0) {
```

Comment:
Also it looks like inside this while(true) we are doing 3 different things. Do you think is possible to use methods for each thing? I think that would help understand better this code. Perhaps validate(), encode() and transformSurrogatePair()